ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp

```diff
@@ -15,72 +15,50 @@
 //*****************************************************************************
 
 #pragma once
-
-#include <cmath>
-
-#include "ngraph/coordinate_transform.hpp"
-#include "ngraph/shape_util.hpp"
-
 namespace ngraph
 {
     namespace runtime
     {
         namespace reference
         {
-            template <typename INDICES_TYPE, typename OUTPUT_TYPE>
-            void one_hot(const INDICES_TYPE* arg,
-                         OUTPUT_TYPE* out,
-                         const Shape& in_shape,
-                         const Shape& out_shape,
-                         size_t one_hot_axis,
-                         const OUTPUT_TYPE on_value,
-                         const OUTPUT_TYPE off_value)
+            template <typename INPUT_TYPE>
+            void one_hot(const INPUT_TYPE* indices,
+                         const Shape& indices_shape,
+                         char* out,
+                         const size_t out_elem_size,
+                         const size_t depth,
```

Comment:
1) A proper output shape is set in [validate_and_infer_types()](https://github.com/openvinotoolkit/openvino/blob/0d2cfdb866e40398ef037abadae570fc5d66b8f8/ngraph/core/src/op/one_hot.cpp#L107-L110). So if we assume that `validate_and_infer_types()` was called before `op::v1::OneHot::evaluate(...)`, it seems that we don't have to pass the whole output shape (which is exactly the input shape + depth inserted at the axis position). Or you believe it is better to re-check the shapes compliance inside the `reference::one_hot<...>(...)`? Like `( shape_size(out_shape) == shape_size(indices_shape) * depth )` for example?
2) Please, note that `( shape_size(out_shape) == shape_size(indices_shape) * depth )` must hold at the time when `reference::one_hot<...>(...)` is called. Therefore `out + (num_ind * depth - 1) * out_elem_size)`  could point outside the allocated output memory block, only if an incorrect `out_elem_size` was provided. More precisely, if `out_elem_size > output_values[0]->get_element_type().size()`. However, this should not be the case, when `reference::one_hot<...>(...)` is called from the [op::v1::OneHot::evaluate(...)](https://github.com/openvinotoolkit/openvino/blob/0d2cfdb866e40398ef037abadae570fc5d66b8f8/ngraph/core/src/op/one_hot.cpp#L142-L149). It is not clear, whether we can easily check that the `out_elem_size` is correct without providing output element type explicitly as a template argument.

Reply:
I think we that we need to have such checks in the evaluate method or here. It should help to avoid any SEGFAULTs for incorrect usages.

Reply:
@ilyachur, agreed, thank you! 
Added a check for I/O shapes consistency:
https://github.com/openvinotoolkit/openvino/blob/3a5eded6af710402ef8de2482af0f03b65313bd5/ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp#L34-L39

The check is performed inside `reference::one_hot<...>(...)`, since potentially it could be called directly through evaluates_map (thus bypassing the `evaluate(...)` method). Please, correct me, if I getting smth wrong here.

Reply:
Moving the checks to the `evaluate(...)` method, since `reference::one_hot<...>(...)` could not be called from evaluates_map according to [this comment](https://github.com/openvinotoolkit/openvino/pull/4330#discussion_r578271336).